### PR TITLE
feat: add flight sql server to expose conclusion feed.

### PR DESCRIPTION
**Pull request by nathanielc**
The Flight SQL Server allows using SQL to query the conclusion feed. Events are read from the db and transformed into conclusion events and returned out the feed.

Closes AES-288

### CODE REVIEW

File: flight/src/server.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+                    DataType::List(Arc::new(Field::new("item", DataType::Binary, false))),
+                    true,
+                ),
+            ])),
+        }
+    }
+    fn highwater_mark_from_expr(expr: &Expr) -> Option<u64> {
+        let find_highwater_mark = |col: &Expr, lit: &Expr| {
+            col.try_as_col()
+                .map(|column| column.name == "index")
```
Comment (dav1do) [inline flight/src/server.rs:104]: nit: `map_or(false, || )` instead of `unwrap_or(false)`

**Review by dav1do — COMMENTED** [review]
Looks good, a couple minor questions. The table provider and execution plan implementations seem reasonable, but I don't have the context to provide much insight there.

File: flight/src/server.rs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+async fn run_service(addr: SocketAddr, svr: Router) -> anyhow::Result<()> {
+    info!(%addr, "FlightSQL server listening");
+    Ok(svr.serve(addr).await?)
+}
+
+/// A ConclusionFeed provides access to [`ConclusionEvent`]s.
+#[async_trait::async_trait]
+pub trait ConclusionFeed: Send + Sync {
+    /// Produce a set of conlcusion events up to the limit with an index greater than the highwater_mark
```
Comment (stbrody) [inline flight/src/server.rs:56]:
```suggestion
    /// Produce a set of conclusion events up to the limit with an index greater than the highwater_mark
```

File: flight/src/types.rs [hunk 3]
```diff
@@ -17,41 +24,58 @@ impl AsRef<ConclusionEvent> for ConclusionEvent {
...
         }
     }
 }
 
-// Dimension is a tuple of (name, value) : way to identify a stream
+/// ConclusionInit is static metadata about a stream.
+/// All events within a stream have the same ConclusionInit.
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct ConclusionInit {
+    /// The CID of the stream. Can be used as a unique identifier for the stream.
```
Comment (stbrody) [inline flight/src/types.rs:39]:
```suggestion
    /// The CID of the init event of the stream. Can be used as a unique identifier for the stream.
```

File: flight/src/types.rs [hunk 4]
```diff
@@ -17,41 +24,58 @@ impl AsRef<ConclusionEvent> for ConclusionEvent {
...
+    /// This is not the StreamId as it does not contain the stream type.
```
Comment (stbrody) [inline flight/src/types.rs:40]:
```suggestion
    /// This is not the StreamId as it does not contain the StreamType.
```

File: flight/src/types.rs [hunk 5]
```diff
@@ -17,41 +24,58 @@ impl AsRef<ConclusionEvent> for ConclusionEvent {
...
+    /// Index of the event. See [`ConclusionEvent`] for invariants about the index.
+    pub index: u64,
+    /// The CID of the event itself. Can be used as a unique identifier for the event.
     pub event_cid: Cid,
+    /// The stream metadata of the event.
     pub init: ConclusionInit,
-    // TODO : rethink this, add a checkpoint to make it work in datafusion query
+    /// Ordered list of previous events this event references.
     pub previous: Vec<Cid>,
+    /// Raw bytes of the event data.
```
Comment (stbrody) [inline flight/src/types.rs:61]: maybe add some information about how this is encoded?

File: one/src/daemon.rs [hunk 6]
```diff
@@ -168,6 +168,14 @@ pub struct DaemonOpts {
         requires = "experimental_features"
     )]
     event_validation: bool,
+
+    /// Flight SQL bind address; Requires using the experimental-features flag
```
Comment (stbrody) [inline one/src/daemon.rs:172]: do we want a separate flag for setting the port number?

**Review by stbrody — COMMENTED** [review]
lgtm, a few small comment suggestions

File: one/src/daemon.rs (hunk 6 above)
Comment (nathanielc) [reply one/src/daemon.rs:172]: I prefer fewer flags, this handle both host and port in a single flag.

File: flight/tests/server.rs [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    FlightSqlServiceClient::new(channel(&addr).await)
+}
+
+async fn execute_flight(
+    client: &mut FlightSqlServiceClient<Channel>,
+    info: FlightInfo,
+) -> Result<RecordBatch> {
+    let schema = Arc::new(Schema::try_from(info.clone()).context("valid schema")?);
+    let mut batches = Vec::with_capacity(info.endpoint.len() + 1);
+    batches.push(RecordBatch::new_empty(schema.clone()));
```
Comment (samika98) [inline flight/tests/server.rs:50]: why are we pushing an empty record batch here. Is it a placeholder for the schema? Can we add a comment here that explains the reasoning.

File: flight/src/server.rs [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> datafusion::common::Result<Arc<dyn ExecutionPlan>> {
+        let schema = projection
+            .map(|projection| self.schema.project(projection))
+            .transpose()?
+            .map(Arc::new)
+            .unwrap_or_else(|| self.schema.clone());
+        debug!(?schema, "projected schema");
+        Ok(Arc::new(FeedExec {
```
Comment (samika98) [inline flight/src/server.rs:160]: nit : I see that the FeedExec implements the ExecutionPlan. Can we add comments as to when the execute gets invoked. Also some explanatory comments in the execute method might help. Why are we setting the limit to 10000 by default?

**Review by samika98 — APPROVED** [review]
Some nit's about more explanatory comments

**Review by dav1do — APPROVED** [review]

File: flight/tests/server.rs (hunk 7 above)
Comment (nathanielc) [reply flight/tests/server.rs:50]: This is test code, the empty batch ensures that when we print out the results the headers always print even if the results are empty. I'll add a comment

File: flight/src/server.rs (hunk 8 above)
Comment (nathanielc) [reply flight/src/server.rs:160]: I'll add some comments, but these type implement the traits from datafusion which are well commented. I recommend reading the TableProvider and ExecutionPlan trait comments.

**Pull request merged**